src/fs/v3/PublicRootWasm.ts

```diff
@@ -0,0 +1,370 @@
+import * as uint8arrays from "uint8arrays"
+import { CID } from "multiformats"
+import { IPFS } from "ipfs-core-types"
+import { initSync, PublicDirectory, PublicFile, PublicNode } from "wnfs"
+
+import { WASM_WNFS_VERSION } from "../../common/version.js"
+import { setup as setupInternal } from "../../setup/internal.js"
+import { FileContent } from "../../ipfs/index.js"
+import { Path } from "../../path.js"
+import { AddResult } from "../../ipfs/index.js"
+import { UnixTree, Puttable, File, Links, PuttableUnixTree } from "../types.js"
+import { BlockStore, IpfsBlockStore } from "./IpfsBlockStore.js"
+import { normalizeFileContent } from "../protocol/public/index.js"
+import { BaseFile } from "../base/file.js"
+import { Metadata } from "../metadata.js"
+
+let initialized = false
+async function initOnce() {
+  if (!initialized) {
+    initialized = true
+    initSync(await setupInternal.wnfsWasmLookup(WASM_WNFS_VERSION))
+  }
+}
+
+interface DirEntry {
+  name: string
+  metadata: {
+    version: "3.0.0"
+    unixMeta: {
+      created: number
+      modified: number
+      mode: number
+      kind: "raw" | "dir" | "file" | "metadata" | "symlink" | "hamtShard"
+    }
+  }
+}
+
+interface OpResult<A> {
+  rootDir: PublicDirectory
+  result: A
+}
+
+export class PublicRootWasm implements UnixTree, Puttable {
+
+  root: Promise<PublicDirectory>
+  lastRoot: PublicDirectory
+  store: BlockStore
+  ipfs: IPFS
+  readOnly: boolean
+
+  constructor(root: PublicDirectory, store: BlockStore, ipfs: IPFS, readOnly: boolean) {
+    this.root = Promise.resolve(root)
+    this.lastRoot = root
+    this.store = store
+    this.ipfs = ipfs
+    this.readOnly = readOnly
+  }
+
+  static async empty(ipfs: IPFS): Promise<PublicRootWasm> {
+    await initOnce()
+    const store = new IpfsBlockStore(ipfs)
+    const root = new PublicDirectory(new Date())
+    return new PublicRootWasm(root, store, ipfs, false)
+  }
+
+  static async fromCID(ipfs: IPFS, cid: CID): Promise<PublicRootWasm> {
+    await initOnce()
+    const store = new IpfsBlockStore(ipfs)
+    const root = await PublicDirectory.load(cid.bytes, store)
+    return new PublicRootWasm(root, store, ipfs, false)
+  }
+
+  private async atomically(fn: (root: PublicDirectory) => Promise<PublicDirectory>) {
+    const root = await this.root
+    this.root = fn(root)
+    await this.root
+  }
+
+  async ls(path: Path): Promise<Links> {
+    const root = await this.root
+    const { result: entries } = await root.ls(path, this.store) as OpResult<DirEntry[]>
+    const result: Links = {}
+    for (const entry of entries) {
+      result[entry.name] = {
+        name: entry.name,
+        isFile: entry.metadata.unixMeta.kind === "file",
+        size: 0, // TODO size?
+        cid: "not provided for performance ", // TODO do we really need a CID here?
```

Comment:
I think Drive depends on this. Why is this a performance win?

Reply:
This requires us to serialize the directory entries in order to hash them. If we always provide a CID, we will always force a serialization.

I guess we should rather include it than leaving it out!

Reply:
I should really change the rs-wnfs `ls` API back from returning metadata to just returning the CIDs of nodes.

Reply:
In the mean time I'm making this get the CID of entries from rs-wnfs :+1:

Reply:
Oh, and I just tested drive, apparently it doesn't depend on that? It works fine like that. Will still change it.

Reply:
Ah nice 👍 Can it still copy the CID as well? (via the context menu of a public item)

Reply:
Yep! The listing in drive actually uses `ipfs.files.stat` on the pretty tree.

While I was looking at the code I also saw some usage of `fs.get` in `resolveItem` that refers to the `cid`, but I think that logic isn't used anywhere. And if it is, it's faulty:

```js
const cid = resolved.cid || resolved.header.content || resolved.header.bareNameFilter
```

The `bareNameFilter` shouldn't be a replacement for a cid.

---

In any case, I still went ahead and fixed this :+1: